### ISSUE TITLE
MicroProfile Fault Tolerance 3.0 (part of MP 4.0)

### ISSUE THREAD

**Issue by Emily-Jiang** [OpenLiberty/open-liberty#11017]
In the upcoming MicroProfile [Fault Tolerance 3.0](https://github.com/eclipse/microprofile-fault-tolerance/milestone/9), we are going to realign FT metrics support with the Metrics 2.0 tag, so we will need to rename metrics names.

When ready, add links to the Upcoming Feature Overview document and Feature Test Summary issue:
- UFO: https://ibm.box.com/s/a75kmfbz89wjch3r0wrj4zhf7xty2ge1
- FTS: #13335
- Beta blog post: #13336
- GA blog post: #15255 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No UFO Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([Adam Yoho](https://github.com/ayoho), [Martin Holder](https://github.com/mhldr), or [Dave Waddling](https://github.com/dave-waddling)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **ID** - ([Kareen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [ ] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by frowe** [comment]
UFO socialization on 7/29/2020 comments/notes:
-Update slide #8 second bullet to indicate beans at all scopes 
-Kevin asked whether we would keep existing metric names and the new ones, discussed and agreed to use only new ones as we have in the past 
Slide 10 - check to see if spec uses the term "scope" or "registry"
Slide 11 - Don noted that with existing metrics, if request-scoped and annotated, the interceptors are likely being created and destroyed for each request and might be a perf issue that should be investigated.
Need to indicate on all affected slides that Context Propagation integration is still in question.

**Comment by Azquelt** [comment]
I've updated the UFO with the above points, except that I've removed all references to Context Propagation since we've now agreed to leave it out of the next version of the spec. Box retains the file history if we need to get that information back in future.

**Comment by Azquelt** [comment]
@steven1046 This is ready for accessibility approval.

The "user interface" of this feature consists of:
* APIs which the user can use in their code
* Metrics exported via MP Metrics

There is no graphical user interface.

**Comment by Azquelt** [comment]
@samwatibm This is ready for globalization approval.

**Comment by samwatibm** [comment]
@Azquelt 2100x translations aren't back yet.   Can you clarify whether there are any translations or whether they were sent previously?   Thanks.

**Comment by Azquelt** [comment]
@samwatibm Yep, most recent change to our message bundles was in #10392 (which was actually a bug fix) with the translations being merged in #15104.

I also had a check for any untranslated messages in our code and didn't find any. Reviewing the content of this feature, I don't think we actually needed to add any new messages.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by Azquelt** [comment]
Thanks @donbourne I'll fill this out.

How should the Metrics reference list be updated? At the moment, it only has metrics from mpMetrics-2.3 and mpFaultTolerance-2.1 in there. The metrics produced by mpFaultTolerance-3.0 are different from those produced by mpFaultTolerance-2.1. Should the metrics reference list include both sets, or should the mpFaultTolerance-2.1 metrics be removed?

**Comment by donbourne** [comment]
@Azquelt , when metrics names changed from mpMetrics-1.x to mpMetrics-2.x we created this doc https://openliberty.io/docs/20.0.0.12/reference/metrics-1-dif.html 
I see you guys have another doc on FT there too about differences between FT-1.x and 2.0... https://openliberty.io/docs/20.0.0.12/reference/fault-tolerance-1-dif.html 

so I think we should make a doc, "Differences between MicroProfile Fault Tolerance 2.x and 3.0" that shows a table of the mapping of metric names.  In the official metrics reference list topic at https://openliberty.io/docs/20.0.0.12/metrics-list.html we would just show the latest metrics from FT 3.0.

**Comment by ayoho** [comment]
FYI: I've removed the FAT approval for now since the feature is not yet using the final version of the MicroProfile API: https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.faulttolerance-3.0.feature#L6. Once the final API is made available and the feature has been updated to use it, I can re-add the approval (also assuming all functionality remains in good working order).

**Comment by Azquelt** [comment]
@ayoho it is actually using the final version of the API: https://github.com/OpenLiberty/open-liberty/blob/integration/dev/cnf/oss_dependencies.maven#L358 https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.org.eclipse.microprofile/faulttolerance.3.0.bnd#L21

However, I did forget to update the maven central reference in the feature file (which I think is used by our maven and gradle plugins to add the correct dependencies when you depend on a feature directly). I'll do that now.

**Comment by Azquelt** [comment]
@donbourne 

**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

    **No** - though this information is included in the STE

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

    **We have not explicitly demoed problem paths for Fault Tolerance 3.0. However, this feature adds incremental improvements over Fault Tolerance 2.1, and the expected problem paths are the same.**

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature? **I think this will be @hanczaryk**
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

    **WL3CDI**

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
    
    **Yes, this feature updates metrics. Docs issues have been raised to update the metrics reference list:**
     - OpenLiberty/docs#3438
     - OpenLiberty/docs#3440

**Comment by Azquelt** [comment]
**ID Summary**
 - Updates to the Open Liberty docs: OpenLiberty/docs#3414
 - Updates to the metrics reference list: OpenLiberty/docs#3438
 - Addition of a page detailing metrics changes in 3.0: OpenLiberty/docs#3440
 - Larger improvements to the docs (not required for this release): OpenLiberty/docs#3293

**Comment by Azquelt** [comment]
@skasund I have completed and uploaded the STE slides here: https://ibm.ent.box.com/file/752343629187 (internal link)

**Comment by skasund** [comment]
@Azquelt  Thanks for the slides.  I've provided the STE approval.

**Comment by donbourne** [comment]
given the comment "incremental improvements over Fault Tolerance 2.1, and the expected problem paths are the same." I think answers look sufficient for serviceability approval.  I will wait for SVT's assessment before approving.

**Comment by hanczaryk** [comment]
Here are the SVT answers for https://github.com/OpenLiberty/open-liberty/issues/11017#issuecomment-744484119

a) Who conducted SVT tests for this feature?  
Brian Hanczaryk

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
Yes, I agree that the serviceability of any problems encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without the need to engage L3.

**Comment by chirp1** [comment]
@Azquelt Hi Andrew, thanks for you comment further up in this epic about the ID work. The one thing in the list that doesn't belong to ID is the #555 guide updates. Gilbert Kwan manages the guides. Would you see him about the updates? I'm approving the epic.
